File: python/progress.py
import os
import json

PROGRESS_PATH = os.path.join(os.path.dirname(__file__), 'progress_vael.json')

PROGRESS = {
    'xp': 0,
    'level': 0,
    'unlocked': ['plains'],
    'towers': {
        'archer': True, 'fire': True, 'ice': False, 'dwarf': True,
        'crossbow': False, 'venom': False, 'druid': False, 'tesla': False,
        'knight': False, 'sniper': False, 'holy': False, 'banner': False, 'warlock': False,
    },
    'maxWaveBeaten': 0,
    'conquestUnlocked': False,
    'completed': False,
    'achievements': {},
}

ACHIEVEMENTS = [
    {'id': 'first_blood', 'name': 'Primera Sangre', 'desc': 'Mata tu primer enemigo', 'icon': '\U0001f5e1\ufe0f'},
    {'id': 'wave_5', 'name': 'Oleada 5', 'desc': 'Sobrevive 5 oleadas', 'icon': '\u2694\ufe0f'},
    {'id': 'wave_10', 'name': 'Veterano', 'desc': 'Sobrevive 10 oleadas', 'icon': '\U0001f6e1\ufe0f'},
    {'id': 'wave_20', 'name': 'Invicto', 'desc': 'Completa todas las 20 oleadas', 'icon': '\U0001f451'},
    {'id': 'all_maps', 'name': 'Explorador', 'desc': 'Desbloquea todos los mapas', 'icon': '\U0001f5fa\ufe0f'},
    {'id': 'gold_500', 'name': 'Avaricia', 'desc': 'Acumula 500 de oro en una partida', 'icon': '\U0001f4b0'},
    {'id': 'kill_100', 'name': 'Carnicero', 'desc': 'Mata 100 enemigos en total', 'icon': '\U0001f480'},
    {'id': 'kill_500', 'name': 'Destructor', 'desc': 'Mata 500 enemigos en total', 'icon': '\U0001f525'},
    {'id': 'hard_win', 'name': 'Masoquista', 'desc': 'Gana en Dificil o superior', 'icon': '\U0001f31f'},
    {'id': 'tower_max', 'name': 'Maestro Constructor', 'desc': 'Mejora una torre al maximo nivel', 'icon': '\U0001f3f7\ufe0f'},
    {'id': 'splitter_kill', 'name': 'Atomizador', 'desc': 'Mata un Escindido y todos sus fragmentos', 'icon': '\U0001f4a5'},
    {'id': 'barracks', 'name': 'General', 'desc': 'Coloca una torre Barracas', 'icon': '\U0001f3d8\ufe0f'},
]

_XP_TABLE = [0, 40, 100, 220, 400, 700, 1100, 1800, 2800, 4500, 7000]

def _load():
    try:
        if os.path.exists(PROGRESS_PATH):
            with open(PROGRESS_PATH, 'r') as f:
                data = json.load(f)
                for k, v in data.items():
                    PROGRESS[k] = v
    except Exception:
        pass

def _save():
    try:
        with open(PROGRESS_PATH, 'w') as f:
            json.dump(PROGRESS, f, indent=2)
    except Exception:
        pass

def progressLoad():
    _load()
    return PROGRESS

def progressSave():
    _save()

def progressGainXp(amount):
    PROGRESS['xp'] += amount
    leveled = False
    while PROGRESS['level'] < len(_XP_TABLE) - 1 and PROGRESS['xp'] >= _XP_TABLE[PROGRESS['level'] + 1]:
        PROGRESS['level'] += 1
        leveled = True
        tower = _unlockTowerForLevel(PROGRESS['level'])
        if tower:
            PROGRESS['towers'][tower] = True
    return leveled

def progressAddMaxWave(n):
    if n > PROGRESS['maxWaveBeaten']:
        PROGRESS['maxWaveBeaten'] = n
    if PROGRESS['maxWaveBeaten'] >= 10:
        PROGRESS['conquestUnlocked'] = True
    _save()

def progressCompleteMap(map_id):
    MAP_ORDER = ['plains', 'desert', 'forest', 'frozen', 'void']
    if map_id not in PROGRESS['unlocked']:
        PROGRESS['unlocked'].append(map_id)
    idx = MAP_ORDER.index(map_id) if map_id in MAP_ORDER else -1
    if 0 <= idx < len(MAP_ORDER) - 1:
        nxt = MAP_ORDER[idx + 1]
        if nxt not in PROGRESS['unlocked']:
            PROGRESS['unlocked'].append(nxt)
    progressGainXp(50)
    _save()

def progressReset():
    PROGRESS['xp'] = 0
    PROGRESS['level'] = 0
    PROGRESS['unlocked'] = ['plains']
    PROGRESS['towers'] = {
        'archer': True, 'fire': True, 'ice': False, 'dwarf': True,
        'crossbow': False, 'venom': False, 'druid': False, 'tesla': False,
        'knight': False, 'sniper': False, 'holy': False, 'banner': False, 'warlock': False,
    }
    PROGRESS['maxWaveBeaten'] = 0
    PROGRESS['conquestUnlocked'] = False
    PROGRESS['completed'] = False
    PROGRESS['achievements'] = {}
    _save()


def unlockAchievement(aid):
    if 'achievements' not in PROGRESS:
        PROGRESS['achievements'] = {}
    if PROGRESS['achievements'].get(aid):
        return False
    PROGRESS['achievements'][aid] = True
    _save()
    return True


def isAchievementUnlocked(aid):
    return PROGRESS.get('achievements', {}).get(aid, False)


def achievementCount():
    return sum(1 for a in ACHIEVEMENTS if isAchievementUnlocked(a['id']))

def _unlockTowerForLevel(level):
    mapping = {
        2: 'ice', 3: 'crossbow', 4: 'venom', 5: 'druid',
        6: 'tesla', 7: 'knight', 8: 'sniper', 9: 'holy',
        10: 'banner', 11: 'warlock',
    }
    return mapping.get(level)

def progressNextMap(last_map_id):
    order = ['plains', 'desert', 'forest', 'frozen', 'void']
    idx = order.index(last_map_id) if last_map_id in order else -1
    if idx < len(order) - 1:
        nxt = order[idx + 1]
        if nxt in PROGRESS['unlocked']:
            return nxt
    return None

_load()

File: js/progress.js
'use strict';

// ============================================================
//  PROGRESIÓN PERSISTENTE DE VALDRYN
//  Guarda el avance del jugador: mapas superados, nivel,
//  experiencia y récords. Se guarda en localStorage.
// ============================================================

var ACHIEVEMENTS = [
  { id: 'first_blood', name: 'Primera Sangre', desc: 'Mata tu primer enemigo', icon: '🗡️' },
  { id: 'wave_5', name: 'Oleada 5', desc: 'Sobrevive 5 oleadas', icon: '⚔️' },
  { id: 'wave_10', name: 'Veterano', desc: 'Sobrevive 10 oleadas', icon: '🛡️' },
  { id: 'wave_20', name: 'Invicto', desc: 'Completa todas las 20 oleadas', icon: '👑' },
  { id: 'all_maps', name: 'Explorador', desc: 'Desbloquea todos los mapas', icon: '🗺️' },
  { id: 'gold_500', name: 'Avaricia', desc: 'Acumula 500 de oro en una partida', icon: '💰' },
  { id: 'kill_100', name: 'Carnicero', desc: 'Mata 100 enemigos en total', icon: '💀' },
  { id: 'kill_500', name: 'Destructor', desc: 'Mata 500 enemigos en total', icon: '🔥' },
  { id: 'hard_win', name: 'Masoquista', desc: 'Gana en Difícil o superior', icon: '🌟' },
  { id: 'tower_max', name: 'Maestro Constructor', desc: 'Mejora una torre al máximo nivel', icon: '🏷️' },
  { id: 'splitter_kill', name: 'Atomizador', desc: 'Mata un Escindido y todos sus fragmentos', icon: '💥' },
  { id: 'barracks', name: 'General', desc: 'Coloca una torre Barracas', icon: '🏰' }
];

var PROGRESS = {
  KEY: 'vaeldryn_save_v2',
  data: null,

  load: function () {
    try {
      this.data = JSON.parse(localStorage.getItem(this.KEY));
    } catch (e) { this.data = null; }
    if (!this.data || typeof this.data !== 'object' || Array.isArray(this.data)) {
      this.data = { maps: {}, xp: 0, totalWaves: 0, wins: 0, playCount: 0, goldEarned: 0, achievements: {} };
    }
    if (!this.data.maps) this.data.maps = {};
    if (!this.data.achievements) this.data.achievements = {};
  },

  save: function () {
    try { localStorage.setItem(this.KEY, JSON.stringify(this.data)); } catch (e) {}
  },

  reset: function () {
    try { localStorage.removeItem(this.KEY); } catch (e) {}
    this.data = { maps: {}, xp: 0, totalWaves: 0, wins: 0, playCount: 0, goldEarned: 0, achievements: {} };
    this.save();
  },

  get xp() { return this.data.xp || 0; },
  get level() { return Math.floor((this.data.xp || 0) / 100) + 1; },
  get levelPct() { return (this.data.xp || 0) % 100; },
  get totalWaves() { return this.data.totalWaves || 0; },
  get wins() { return this.data.wins || 0; },
  get playCount() { return this.data.playCount || 0; },
  get goldEarned() { return this.data.goldEarned || 0; },

  startGoldBonus: function () {
    return Math.min(80, (this.level - 1) * 5);
  },

  mapProgress: function (id) {
    return this.data.maps[id] || { bestWave: 0, completed: false, stars: 0 };
  },

  isMapUnlocked: function (id) {
    for (var i = 0; i < MAPS.length; i++) {
      if (MAPS[i].id === id) {
        if (i === 0) return true;
        var prev = this.data.maps[MAPS[i - 1].id];
        return !!prev && prev.completed === true;
      }
    }
    return true;
  },

  // ¿la torre está desbloqueada de forma permanente?
  isTowerUnlocked: function (type) {
    var def = TOWERS[type];
    if (!def || !def.unlock) return true;
    if (def.unlock.map) {
      var m = this.data.maps[def.unlock.map];
      return !!m && m.completed === true;
    }
    return true;
  },

  recordRun: function (mapId, game) {
    var m = this.data.maps[mapId] || { bestWave: 0, completed: false, stars: 0, goldEarned: 0 };
    m.bestWave = Math.max(m.bestWave || 0, game.wave);
    m.goldEarned = (m.goldEarned || 0) + Math.round(game.gold * 0.3);
    if (game.wave >= CONFIG.WIN_WAVE) m.completed = true;
    var pct = Math.max(0, Math.min(1, game.wave / CONFIG.WIN_WAVE));
    var stars = pct >= 1 ? 3 : (pct >= 0.7 ? 2 : (pct >= 0.35 ? 1 : 0));
    m.stars = Math.max(m.stars || 0, stars);
    this.data.maps[mapId] = m;
    this.data.totalWaves = Math.max(this.data.totalWaves || 0, game.wave);
    this.data.maxWaveBeaten = Math.max(this.data.maxWaveBeaten || 0, game.wave);
    if ((this.data.maxWaveBeaten || 0) >= 10) this.data.conquestUnlocked = true;
    this.data.goldEarned = (this.data.goldEarned || 0) + game.kills * 4;
    this.data.playCount = (this.data.playCount || 0) + 1;
    var gained = Math.round(game.wave * 6 + game.kills * 0.5);
    this.addXp(gained);
    if (game.wave >= CONFIG.WIN_WAVE) this.data.wins = (this.data.wins || 0) + 1;
    this.save();
    return gained;
  },

  addXp: function (n) {
    this.data.xp = (this.data.xp || 0) + n;
  }
};

function maxWaveBeaten() {
  if (!PROGRESS.data) return 0;
  var best = PROGRESS.data.maxWaveBeaten || 0;
  for (var mk in PROGRESS.data.maps) best = Math.max(best, PROGRESS.data.maps[mk].bestWave || 0);
  return best;
}

function conquestUnlocked() {
  return !!PROGRESS.data.conquestUnlocked || maxWaveBeaten() >= 10;
}

function unlockAchievement(aid) {
  if (!PROGRESS.data.achievements) PROGRESS.data.achievements = {};
  if (PROGRESS.data.achievements[aid]) return false;
  PROGRESS.data.achievements[aid] = true;
  PROGRESS.save();
  return true;
}

function isAchievementUnlocked(aid) {
  return !!(PROGRESS.data && PROGRESS.data.achievements && PROGRESS.data.achievements[aid]);
}

function achievementCount() {
  var n = 0;
  for (var i = 0; i < ACHIEVEMENTS.length; i++) {
    if (isAchievementUnlocked(ACHIEVEMENTS[i].id)) n++;
  }
  return n;
}

// --- Clasificación local (mejores partidas) ---
var SCORES = {
  key: 'vaeldryn_scores',
  list: [],

  load: function () {
    try {
      this.list = JSON.parse(localStorage.getItem(this.key) || '[]');
    } catch (e) { this.list = []; }
    if (!Array.isArray(this.list)) this.list = [];
    return this.list;
  },

  save: function () {
    try { localStorage.setItem(this.key, JSON.stringify(this.list)); } catch (e) { /* ignorar */ }
  },

  add: function (entry) {
    entry.date = Date.now();
    this.list.push(entry);
    this.list.sort(function (a, b) {
      return (b.wave - a.wave) || (b.kills - a.kills);
    });
    if (this.list.length > 10) this.list.length = 10;
    this.save();
    return this.list.indexOf(entry);
  }
};
